### PR TITLE
Added REST Message / Method resources

### PR DESCRIPTION
**Pull request by mikegron**
Added resources to provider:

- servicenow_rest_message
- servicenow_rest_message_header
- servicenow_rest_method
- servicenow_rest_method_header

### CODE REVIEW

File: servicenow/client/client_rest_message.go [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+
+import (
+	"fmt"
+)
+
+const endpointRestMessage = "sys_rest_message.do"
+
+// RestMessage represents the json response for a REST Message in ServiceNow.
+type RestMessage struct {
+	BaseResult
```
Comment (erocheleau) [inline servicenow/client/client_rest_message.go:11]: Comments?

File: servicenow/client/client_rest_message.go [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+	AuthenticationType string `json:"authentication_type"`
+}
+
+// RestMessageResults is the object returned by ServiceNow API when saving or retrieving records.
+type RestMessageResults struct {
+	Records []RestMessage `json:"records"`
+}
+
+// GetRestMessage retrieves a specific REST Message in ServiceNow with it's sys_id.
+func (client *ServiceNowClient) GetRestMessage(id string) (*RestMessage, error) {
```
Comment (erocheleau) [inline servicenow/client/client_rest_message.go:25]: Are those already defined somewhere as an Interface?

```
type ServiceNowClient interface {
  GetRestMessage(id String) (*RestMessage, error)
  CreateRestMessage(restMessage *RestMessage) (*RestMessage, error)
  UpdateRestMessage(restMessage *RestMessage) error
  DeleteRestMessage(id string) error
}

File: servicenow/client/client_rest_message.go [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+func (client *ServiceNowClient) DeleteRestMessage(id string) error {
+	return client.deleteObject(endpointRestMessage, id)
+}
+
+func (results RestMessageResults) validate() error {
+	if len(results.Records) <= 0 {
+		return fmt.Errorf("no records found")
+	} else if len(results.Records) > 1 {
+		return fmt.Errorf("more than one record received")
+	} else if results.Records[0].Status != "success" {
```
Comment (erocheleau) [inline servicenow/client/client_rest_message.go:59]: Worth making constants for the error messages?

File: servicenow/client/client_rest_method.go [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+		return nil, err
+	}
+
+	return &restMethodResults.Records[0], nil
+}
+
+// CreateRestMethod creates a HTTP method in ServiceNow and returns the newly created JS Include.
+func (client *ServiceNowClient) CreateRestMethod(RestMethod *RestMethod) (*RestMethod, error) {
+	restMethodResults := RestMethodResults{}
+	if err := client.createObject(endpointRestMethod, RestMethod.Scope, RestMethod, &restMethodResults); err != nil {
```
Comment (erocheleau) [inline servicenow/client/client_rest_method.go:38]: I don't see Scope as an attribute of RestMethod struct is that normal?

File: servicenow/client/client_rest_message.go [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+		return nil, err
+	}
+
+	return &restMessageResults.Records[0], nil
+}
+
+// CreateRestMessage creates a REST Message in ServiceNow and returns the newly created endpoint.
+func (client *ServiceNowClient) CreateRestMessage(restMessage *RestMessage) (*RestMessage, error) {
+	restMessageResults := RestMessageResults{}
+	if err := client.createObject(endpointRestMessage, restMessage.Scope, restMessage, &restMessageResults); err != nil {
```
Comment (erocheleau) [inline servicenow/client/client_rest_message.go:37]: I don't see Scope anywhere here on RestMessage.

File: servicenow/client/client_rest_method_header.go [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+		return nil, err
+	}
+
+	return &restMethodHeaderResults.Records[0], nil
+}
+
+// CreateRestMethodHeader creates a HTTP header in ServiceNow and returns the newly created JS Include.
+func (client *ServiceNowClient) CreateRestMethodHeader(restMethodHeader *RestMethodHeader) (*RestMethodHeader, error) {
+	restMethodHeaderResults := RestMethodHeaderResults{}
+	if err := client.createObject(endpointRestMethodHeader, restMethodHeader.Scope, restMethodHeader, &restMethodHeaderResults); err != nil {
```
Comment (erocheleau) [inline servicenow/client/client_rest_method_header.go:35]: Scope again?

File: servicenow/client/client_rest_method_header.go [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+func (client *ServiceNowClient) UpdateRestMethodHeader(restMethodHeader *RestMethodHeader) error {
+	return client.updateObject(endpointRestMethodHeader, restMethodHeader.Id, restMethodHeader)
+}
+
+// DeleteRestMethodHeader deletes a HTTP header in ServiceNow with the corresponding sys_id.
+func (client *ServiceNowClient) DeleteRestMethodHeader(id string) error {
+	return client.deleteObject(endpointRestMethodHeader, id)
+}
+
+func (results RestMethodHeaderResults) validate() error {
```
Comment (erocheleau) [inline servicenow/client/client_rest_method_header.go:52]: This validate method is repeated like 3 times shouldn't these kind of things be "extended" from a base interface?

File: servicenow/resources/resource_rest_message.go [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+import (
+	"github.com/coveo/terraform-provider-servicenow/servicenow/client"
+	"github.com/hashicorp/terraform/helper/schema"
+)
+
+const restMessageName = "name"
+const restMessageDescription = "description"
+const restMessageRestEndpoint = "rest_endpoint"
+const restMessageAccess = "access"
+const restMessageAuthenticationType = "authentication_type" // No auth is supported
```
Comment (erocheleau) [inline servicenow/resources/resource_rest_message.go:12]: What does that mean? No auth is supported?

**Review by erocheleau — APPROVED** [review]
Other than what looks like a lot of repeated code and little changes it looks fine

File: servicenow/client/client_rest_method_header.go (hunk 7 above)
Comment (mikegron) [reply servicenow/client/client_rest_method_header.go:52]: I'm taking notes and will refactor everything afterwards :)

File: servicenow/client/client_rest_message.go (hunk 5 above)
Comment (mikegron) [reply servicenow/client/client_rest_message.go:37]: It's in the BaseResult struct

File: servicenow/client/client_rest_message.go (hunk 2 above)
Comment (mikegron) [reply servicenow/client/client_rest_message.go:25]: negative

**Review by erocheleau — APPROVED** [review]

**Pull request merged**